src/OpenTelemetry.Instrumentation.Dependencies/Implementation/GrpcClientDiagnosticListener.cs

```diff
@@ -0,0 +1,124 @@
+                return;
+            }
+
+            var grpcMethodTag = activity.Tags.FirstOrDefault(tag => tag.Key == "grpc.method");
+            var grpcMethod = grpcMethodTag.Value?.Trim('/');
+
+            this.Tracer.StartActiveSpanFromActivity(grpcMethod, activity, SpanKind.Client, out var span);
+
+            if (span.IsRecording)
+            {
+                var rpcService = GrpcMethodRegex.Match(grpcMethod).Groups["service"].Value;
+
+                span.SetAttribute("rpc.service", rpcService);
+
+                var uriHostNameType = Uri.CheckHostName(request.RequestUri.Host);
+                if (uriHostNameType == UriHostNameType.IPv4 || uriHostNameType == UriHostNameType.IPv6)
+                {
+                    span.SetAttribute("net.peer.ip", request.RequestUri.Host);
+                }
+                else
+                {
+                    span.SetAttribute("net.peer.name", request.RequestUri.Host);
+                }
+
+                span.SetAttribute("net.peer.port", request.RequestUri.Port);
+            }
+
+            // Do we want to propagate trace context here? GrpcClient is built on top of HttpClient.
+            // Currently a span is generated by both GrpcClient and HttpClient. This is probably redundant.
+            // If we do propagation here, then the HttpClient instrumentation would be skipped as it checks if a traceparent header is already present
+        }
+
+        public override void OnStopActivity(Activity activity, object payload)
+        {
+            const string EventNameSuffix = ".OnStopActivity";
+            var span = this.Tracer.CurrentSpan;
+            try
+            {
+                if (span == null || !span.Context.IsValid)
+                {
+                    InstrumentationEventSource.Log.NullOrBlankSpan(nameof(GrpcClientDiagnosticListener) + EventNameSuffix);
+                    return;
+                }
+
+                if (span.IsRecording)
+                {
+                    span.Status = Status.Unknown;
+
+                    var grpcStatusCodeTag = activity.Tags.FirstOrDefault(tag => tag.Key == "grpc.status_code").Value;
+                    if (int.TryParse(grpcStatusCodeTag, out var statusCode))
+                    {
+                        span.PutRpcStatusCode(statusCode);
+                    }
+                }
+            }
+            finally
+            {
+                span?.End();
+            }
+        }
+
+        public override void OnException(Activity activity, object payload)
+        {
+            const string EventNameSuffix = ".OnException";
+            var span = this.Tracer.CurrentSpan;
+
+            if (span == null || !span.Context.IsValid)
+            {
+                InstrumentationEventSource.Log.NullOrBlankSpan(nameof(GrpcClientDiagnosticListener) + EventNameSuffix);
+                return;
+            }
+
+            // set span.Status
+            // validate whether we need to end span here
```

Comment:
Does more need to be done here?

Reply:
Apologies for the delay... yes, I still have a bit to do. Not quite ready for approval.

Reply:
Oh, sorry, I didn't quite realize what I was looking at. This code has been removed since an `OnException` event is not a thing for the GrpcClient.